### PR TITLE
Adding Matcher for Transaction Filters

### PR DESCRIPTION
**Pull request by barnjamin**
Most contracts do some basic transaction matching at the beginning to make sure the incoming transaction group is well formed. This abstracts the pattern and provides a convenient way to mix in common fields to check.

This does not allow txn field arrays yet and no nice api to refer to other transactions in the group

### CODE REVIEW

**Review by algochoi — COMMENTED** [review]
This looks really useful in terms of reducing boilerplate code 😃 

My initial two cents: I'm not too sure about using dicts for the matcher in case people accidentally add weird fields in there. I don't know the exact design details or motivation of this, but maybe it could look something a little more restrictive like:

```
class Common:
   # initializers 
   # could just wrap the dictionary in this class or translate the fields into a key-value in the Match class?
class PaymentMatch(Common):
   # have some functions specific to payments or initialize certain params
   def __init__(amt, sender, ...):
       # ...
```

Then the `payMatcher` could look something like:
```
payMatcher = PaymentMatch(amt=100)
```
which looks a little more readable to my eyes and hopefully worry less about overriding/injecting random keys in the dict. Or we could just keep the dictionary and have a method that just constructs a `payMatcher` or `axferMatcher`. Anyway, looks promising.

File: pytealutils/transaction/test_match.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
+from pyteal import Approve, Arg, Bytes, Cond, Gtxn, Return
+from pytest import *
+
+from tests.helpers import *
+
+from .match import *
```
Comment (tzaffi) [inline pytealutils/transaction/test_match.py:6]:
```suggestion
from .match import AssetIdAndAmount, Common, HasMinFee, Match, PaymentAmount
```

File: pytealutils/transaction/test_match.py [hunk 2]
```diff
@@ -0,0 +1,41 @@
+from pyteal import Approve, Arg, Bytes, Cond, Gtxn, Return
```
Comment (tzaffi) [inline pytealutils/transaction/test_match.py:1]:
```suggestion
from pyteal import Approve, Arg, Bytes, Cond, Gtxn, Return, TxnField
```

File: pytealutils/transaction/test_match.py [hunk 3]
```diff
@@ -0,0 +1,41 @@
+from pyteal import Approve, Arg, Bytes, Cond, Gtxn, Return, TxnField
+from pytest import *
```
Comment (tzaffi) [inline pytealutils/transaction/test_match.py:2]:
```suggestion
```

File: pytealutils/transaction/match.py [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+    TxnField,
+    TxnType,
+)
+
+# Common field checks
+
+NoRekey = {TxnField.rekey_to: Global.zero_address()}
+NoCloseTo = {TxnField.close_remainder_to: Global.zero_address()}
+NoCloseAsset = {TxnField.asset_close_to: Global.zero_address()}
+Common = {**NoRekey, **NoCloseTo, **NoCloseAsset}
```
Comment (tzaffi) [inline pytealutils/transaction/match.py:22]: Maybe rename something a bit more indicative such as `CommonExclusions` or `NoRekeyOrCloses` ?

File: pytealutils/transaction/match.py [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+
+HasMinFee = {TxnField.fee: Global.min_txn_fee()}
+
+Payment = {TxnField.type_enum: TxnType.Payment}
```
Comment (tzaffi) [inline pytealutils/transaction/match.py:26]: `IsPayment` ?

Comment (tzaffi) [reply pytealutils/transaction/match.py:26]: Hmmm... maybe that's too boolean'ish. So maybe `SetPaymentType`...

**Pull request closed without merging**